xapian-letor/feature/colltfcolllenfeature.cc

```diff
@@ -42,28 +42,26 @@ CollTfCollLenFeature::get_values() const
 
     vector<double> values;
     double value = 0;
-    double coll_len = 0;
-    map<string, Xapian::termcount>::const_iterator coll_len_iterator =
-	    collection_length.find("title");
+    double coll_len;
+    auto coll_len_iterator = collection_length.find("title");
     if (coll_len_iterator != collection_length.end())
 	coll_len = (double)coll_len_iterator->second;
     else
 	coll_len = 0;
 
     for (Xapian::TermIterator qt = feature_query.get_unique_terms_begin();
 	 qt != feature_query.get_terms_end(); ++qt) {
-	if ((*qt).substr(0, 1) == "S" || (*qt).substr(1, 1) == "S") {
-	    map<string, Xapian::termcount>::const_iterator coll_tf_iterator =
-		    collection_termfreq.find(*qt);
+	if (is_title_term((*qt))) {
+	    auto coll_tf_iterator = collection_termfreq.find(*qt);
 	    if (coll_tf_iterator != collection_termfreq.end()) {
 		value += log10(1 + (coll_len /
 				    (double)(1 + coll_tf_iterator->second)));
-	    }
-	    else
+	    } else {
 		value += log10(1 + coll_len);
-	}
-	else
+	    }
+	} else {
 	    value += 0;
```

Comment:
I keep saying to get rid of these pointless `value += 0;` statements!

They don't make much sense to the reader of the code, and the compiler generally can't optimise them away (because IEEE FP supports `-0`, and `(-0) + (+0)` is `+0`.  `+0` is what you'll get for a floating point constant for `0`, so the upshot is that adding `0` to a double changes it in the corner case where the input is `-0`, and `x + 0` can't be optimised to just `x` unless the compiler can prove that `x` can't be `-0`).

Checking recent GCC and clang, neither optimises away the addition for a simple test case:

```
$ cat test.cc
double f(double x) {
   x += 0;
   return x;
}
$ g++ -S -o - -O2 test.cc
	.file	"test.cc"
	.text
	.p2align 4,,15
	.globl	_Z1fd
	.type	_Z1fd, @function
_Z1fd:
.LFB0:
	.cfi_startproc
	addsd	.LC0(%rip), %xmm0
	ret
	.cfi_endproc
.LFE0:
	.size	_Z1fd, .-_Z1fd
	.section	.rodata.cst8,"aM",@progbits,8
	.align 8
.LC0:
	.long	0
	.long	0
	.ident	"GCC: (Debian 6.3.0-18) 6.3.0 20170516"
	.section	.note.GNU-stack,"",@progbits
```

`addsd	.LC0(%rip), %xmm0` loads constant `0.0` from `.LC0` and adds it.

```
$ clang++-4.0 -S -o - -O2 test.cc
	.text
	.file	"test.cc"
	.globl	_Z1fd
	.p2align	4, 0x90
	.type	_Z1fd,@function
_Z1fd:                                  # @_Z1fd
	.cfi_startproc
# BB#0:
	xorpd	%xmm1, %xmm1
	addsd	%xmm1, %xmm0
	retq
.Lfunc_end0:
	.size	_Z1fd, .Lfunc_end0-_Z1fd
	.cfi_endproc


	.ident	"clang version 4.0.1-+rc2-1 (tags/RELEASE_401/rc2)"
	.section	".note.GNU-stack","",@progbits
```

clang XORs a register with itself to generate the constant, but otherwise it's the same.

If you pass `-ffast-math` the compiler is allowed to use non-confirming assumptions when optimising, and the function becomes empty:

```
$ g++ -S -o - -ffast-math -O2 test.cc
	.file	"test.cc"
	.text
	.p2align 4,,15
	.globl	_Z1fd
	.type	_Z1fd, @function
_Z1fd:
.LFB0:
	.cfi_startproc
	rep ret
	.cfi_endproc
.LFE0:
	.size	_Z1fd, .-_Z1fd
	.ident	"GCC: (Debian 6.3.0-18) 6.3.0 20170516"
	.section	.note.GNU-stack,"",@progbits
```